### PR TITLE
store: Add Groupcache as a cache backend

### PR DESCRIPTION
**Pull request by akanshat**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
- Added Groupcache.
- Implemented Cache interface to groupcache.
- Added a new package cachekey.

## Verification
Existing tests passed.

### CODE REVIEW

**Review by GiedriusS — COMMENTED** [review]
Seems like you have some unrelated changes here? Could you please rebase on main and don't forget the DCO ;P

**Comment by akanshat** [comment]
Okay, I was hoping you won't see this. 😆

File: pkg/cache/groupcache.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+func (c *Groupcache) Fetch(ctx context.Context, keys []string) map[string][]byte {
+	data := map[string][]byte{}
+
+	for _, k := range keys {
+		var keyData []byte
+
+		if err := c.group.Get(ctx, k, groupcache.AllocatingByteSliceSink(&keyData)); err != nil {
+			level.Error(c.logger).Log("msg", "failed fetching data from groupcache", "err", err, "key", k)
+			return nil
```
Comment (GiedriusS) [inline pkg/cache/groupcache.go:211]: Who wrote this code?!? :joy: :joy: maybe let's do `continue` here to return partial results if Get() fails?

Comment (akanshat) [reply pkg/cache/groupcache.go:211]: Let's do a git blame 😆

**Review by GiedriusS — COMMENTED** [review]
Looks good, just need to convert to https://github.com/vimeo/galaxycache now + add some e2e tests together with metrics? :smile:

**Comment by akanshat** [comment]
I am planning to make groupcache work properly first, and write a test for it too. Then I want to convert it to galaxycache.

File: pkg/cache/groupcache.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+		backendLoads:      backendLoads,
+		backendLoadErrors: backendLoadErrors,
+		cacheHits:         cacheHits,
+	}
+	reg.MustRegister(collector)
+}
+
+func (s CacheStatsCollector) Collect(ch chan<- prometheus.Metric) {
+	stats := s.galaxy.Stats
+	ch <- prometheus.MustNewConstMetric(s.gets, prometheus.CounterValue, float64(stats.Gets))
```
Comment (GiedriusS) [inline pkg/cache/groupcache.go:272]: I think we need to call `.Get()` everywhere here to avoid races (:

File: pkg/cache/groupcache.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+	ch <- prometheus.MustNewConstMetric(s.loads, prometheus.CounterValue, float64(stats.Loads))
+	ch <- prometheus.MustNewConstMetric(s.peerLoads, prometheus.CounterValue, float64(stats.PeerLoads))
+	ch <- prometheus.MustNewConstMetric(s.peerLoadErrors, prometheus.CounterValue, float64(stats.PeerLoadErrors))
+	ch <- prometheus.MustNewConstMetric(s.backendLoads, prometheus.CounterValue, float64(stats.BackendLoads))
+	ch <- prometheus.MustNewConstMetric(s.backendLoadErrors, prometheus.CounterValue, float64(stats.BackendLoadErrors))
+	ch <- prometheus.MustNewConstMetric(s.cacheHits, prometheus.CounterValue, float64(stats.MaincacheHits), galaxycache.MainCache.String())
+	ch <- prometheus.MustNewConstMetric(s.cacheHits, prometheus.CounterValue, float64(stats.HotcacheHits), galaxycache.HotCache.String())
+}
+
+func (s CacheStatsCollector) Describe(ch chan<- *prometheus.Desc) {
```
Comment (GiedriusS) [inline pkg/cache/groupcache.go:282]: These need to be:
```suggestion
func (s *CacheStatsCollector) Describe(ch chan<- *prometheus.Desc) {
```

Since you are `Register()`ing a pointer to `CacheStatsCollector` :P

File: pkg/cache/groupcache.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+		peerLoads:         peerLoads,
+		peerLoadErrors:    peerLoadErrors,
+		backendLoads:      backendLoads,
+		backendLoadErrors: backendLoadErrors,
+		cacheHits:         cacheHits,
+	}
+	reg.MustRegister(collector)
+}
+
+func (s CacheStatsCollector) Collect(ch chan<- prometheus.Metric) {
```
Comment (GiedriusS) [inline pkg/cache/groupcache.go:270]:
```suggestion
func (s *CacheStatsCollector) Collect(ch chan<- prometheus.Metric) {
```

**Review by GiedriusS — COMMENTED** [review]
After converting those two methods to have a pointer as the receiver, the metrics appear and I now see the following error when running the e2e tests:
```
             unexpected error: unable to find metrics [thanos_cache_groupcache_hits_total] with expected values after 50 retries. Last error: <nil>. Last values: [31]
```

This is probably because Store "refreshed" the view of blocks asynchronously hence increasing that counter (:

**Comment by akanshat** [comment]
@GiedriusS What did you mean by `.Get()`, I couldn't find any such method here.

**Comment by GiedriusS** [comment]
> @GiedriusS What did you mean by `.Get()`, I couldn't find any such method here.

`Get()` is a method on `AtomicInt` which is the type of `stats.Gets` and so on. So, it should be:

```go
	ch <- prometheus.MustNewConstMetric(s.gets, prometheus.CounterValue, float64(stats.Gets.Get()))
``` 

And so on, I think.

**Comment by hanjm** [comment]
Awesome feature! I am headache in lots of memcache timeout log, groupcache will save me 😄

File: pkg/cache/groupcache.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+	))
+
+	// Register GalaxyCache stats.
```
Comment (onprem) [inline pkg/cache/groupcache.go:213]: Maybe we can move this on top of the function itself. And maybe we can reword it a bit to be more detailed.

File: pkg/cache/groupcache.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+		universe: universe,
+	}, nil
+}
+
+func (c *Groupcache) Store(ctx context.Context, data map[string][]byte, ttl time.Duration) {
+	// Noop since cache is already filled during fetching.
+}
+
+func (c *Groupcache) Fetch(ctx context.Context, keys []string) map[string][]byte {
+	data := map[string][]byte{}
```
Comment (onprem) [inline pkg/cache/groupcache.go:228]: TIL that this doesn't create a `nil` map. :open_mouth:

File: test/e2e/store_gateway_test.go [hunk 7]
```diff
@@ -398,3 +398,157 @@ blocks_iter_ttl: 0s`, memcached.InternalEndpoint("memcached"))
...
+				Insecure:  true,
+			},
+		},
+		fmt.Sprintf(groupcacheConfig, 3),
+	)
+
+	testutil.Ok(t, err)
+	testutil.Ok(t, e2e.StartAndWaitReady(store3))
+
+	q, err := e2ethanos.NewQuerierBuilder(e, "1", store1.InternalEndpoint("grpc")).Build()
```
Comment (onprem) [inline test/e2e/store_gateway_test.go:481]: We are missing `store2` and `store3` here.

File: test/e2e/store_gateway_test.go [hunk 8]
```diff
@@ -398,3 +398,157 @@ blocks_iter_ttl: 0s`, memcached.InternalEndpoint("memcached"))
...
+	testutil.Ok(t, objstore.UploadDir(ctx, l, bkt, path.Join(dir, id.String()), id.String()))
+
+	// Wait for store to sync blocks.
+	// thanos_blocks_meta_synced: 1x loadedMeta 0x labelExcludedMeta 0x TooFreshMeta.
+	testutil.Ok(t, store1.WaitSumMetrics(e2e.Equals(1), "thanos_blocks_meta_synced"))
+	testutil.Ok(t, store1.WaitSumMetrics(e2e.Equals(0), "thanos_blocks_meta_sync_failures_total"))
+
+	testutil.Ok(t, store1.WaitSumMetrics(e2e.Equals(1), "thanos_bucket_store_blocks_loaded"))
+	testutil.Ok(t, store1.WaitSumMetrics(e2e.Equals(0), "thanos_bucket_store_block_drops_total"))
+	testutil.Ok(t, store1.WaitSumMetrics(e2e.Equals(0), "thanos_bucket_store_block_load_failures_total"))
```
Comment (onprem) [inline test/e2e/store_gateway_test.go:517]: We should at least check for `thanos_bucket_store_blocks_loaded` to become `1` for `store2` and `store3` as well.

File: test/e2e/store_gateway_test.go [hunk 9]
```diff
@@ -398,3 +398,157 @@ blocks_iter_ttl: 0s`, memcached.InternalEndpoint("memcached"))
...
+				{
+					"a":       "1",
+					"b":       "2",
+					"ext1":    "value1",
+					"replica": "1",
+				},
+			},
+		)
+
+		testutil.Ok(t, store1.WaitSumMetricsWithOptions(e2e.Greater(0), []string{`thanos_cache_groupcache_loads_total`}))
```
Comment (onprem) [inline test/e2e/store_gateway_test.go:534]: Lets check for the generic metrics we check in memcached test here as well.

File: pkg/cache/groupcache.go [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+	// Addresses of statically configured peers (repeatable). The scheme may be prefixed with 'dns+' or 'dnssrv+' to detect store API servers through respective DNS lookups.
+	// Typically, you'd want something like `dns+http://thanos-store:42/`.
+	Peers []string `yaml:"peers"`
+
+	// Address of ourselves in the peer list. This needs to be set to `http://external-ip:HTTP_PORT`
+	// of the current instance.
+	SelfURL string `yaml:"self_url"`
+
+	// Maximum size of the hot in-memory cache.
+	MaxSize model.Bytes `yaml:"max_size"`
```
[the remark this answers is not in the document]
Comment (onprem) [reply pkg/cache/groupcache.go:59]: `model.Bytes` supports unmarshalling with units (for ex. `250M`), so it seems fine as is.

File: pkg/cache/groupcache.go [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+type GroupcacheConfig struct {
+	// Addresses of statically configured peers (repeatable). The scheme may be prefixed with 'dns+' or 'dnssrv+' to detect store API servers through respective DNS lookups.
+	// Typically, you'd want something like `dns+http://thanos-store:42/`.
+	Peers []string `yaml:"peers"`
+
+	// Address of ourselves in the peer list. This needs to be set to `http://external-ip:HTTP_PORT`
+	// of the current instance.
+	SelfURL string `yaml:"self_url"`
+
+	// Maximum size of the hot in-memory cache.
```
Comment (onprem) [inline pkg/cache/groupcache.go:58]:
```suggestion
	// Maximum size of the in-memory cache.
```

File: pkg/cache/groupcache.go [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+				finalAttrs, err := json.Marshal(attrs)
+				if err != nil {
+					return err
+				}
+				err = dest.UnmarshalBinary(finalAttrs)
+				if err != nil {
+					return err
+				}
+			case cachekey.IterVerb:
+				// Not supported.
```
Comment (onprem) [inline pkg/cache/groupcache.go:162]: Lets figure out why we don't support Iter.

File: Dockerfile.multi-stage [hunk 13]
```diff
@@ -1,6 +1,6 @@
 # By default we pin to amd64 sha. Use make docker to automatically adjust for arm64 versions.
 ARG BASE_DOCKER_SHA="14d68ca3d69fceaa6224250c83d81d935c053fb13594c811038c461194599973"
-FROM golang:1.16-alpine3.12 as builder
+FROM golang:1.17-alpine3.15 as builder
```
Comment (GiedriusS) [inline Dockerfile.multi-stage:3]: This change also seems unrelated. Probably just needs a rebase since this exact version is used in `main`

File: CHANGELOG.md [hunk 14]
```diff
@@ -6,7 +6,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) a
 
 NOTE: As semantic versioning states all 0.y.z releases can contain breaking changes in API (flags, grpc API, any backward compatibility)
 
-We use *breaking :warning:* to mark changes that are not backward compatible (relates only to v0.y.z releases.)
+We use _breaking :warning:_ to mark changes that are not backward compatible (relates only to v0.y.z releases.)
```
Comment (GiedriusS) [inline CHANGELOG.md:9]: It seems like CHANGELOG.md has lots of unrelated changes? Let's only add:
```
- [#4818](https://github.com/thanos-io/thanos/pull/4818) _EXPERIMENTAL_ store: Add Groupcache as a cache backend. [Original issue](https://github.com/thanos-io/thanos/issues/2962).
```

**Review by GiedriusS — COMMENTED** [review]
It became a big PR in the end but we've done it in small iterations so it's all good :tada:

**Review by GiedriusS — APPROVED** [review]
:tada: :tada: :tada: 
https://github.com/thanos-io/thanos/issues/5037 made this to track the next steps

**Review by GiedriusS — APPROVED** [review]
Just made some random changes to make the linters happy, I hope you don't mind :+1:

**Comment by akanshat** [comment]
Thanks!

**Comment by GiedriusS** [comment]
Only docs are failing so I'll merge with admin rights :tada:
